### PR TITLE
vtctldclient OnlineDDL CANCEL

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Implementing `vtctldclient OnlineDDL cancel`, as part of https://github.com/vitessio/vitess/issues/13712#event-10007807200.

This PR is generally similar to https://github.com/vitessio/vitess/pull/13830, but with one notable change: support for `"all"` argument. Some commands, such as `CANCEL`, `COMPLETE`, `THROTTLE` support the `ALL` argument on top of a UUID argument. This PR generalizes `alterSingleSchemaMigrationQuery()`, now renamed as `alterSchemaMigrationQuery()`.

Usage for this command:

```sh
$ vtctldclient OnlineDDL cancel commerce 326fb621_44c9_11ee_8cc6_0a43f95f28a3

$ vtctldclient OnlineDDL cancel commerce all
```

## Related Issue(s)

Closes https://github.com/vitessio/vitess/issues/13717

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by shlomi-noach** [comment]
If this is approved, I'll do all the rest (`LAUNCH`, `COMPLETE`, `THROTTLE`, `UNTHROTTLE`) in the exact same way. Postponing documentation PR - thinking I'll just do one PR for all changes instead of 5 different PRs.

File: go/cmd/vtctldclient/command/onlineddl.go [hunk 1]
```diff
@@ -32,13 +32,25 @@ import (
...
 var (
 	OnlineDDL = &cobra.Command{
 		Use:                   "OnlineDDL <cmd> <keyspace> [args]",
 		Short:                 "Operates on online DDL (schema migrations).",
 		DisableFlagsInUseLine: true,
 		Args:                  cobra.MinimumNArgs(2),
 	}
+	OnlineDDLCancel = &cobra.Command{
+		Use:                   "cancel <keyspace> <uuid|all>",
+		Short:                 "CancelSchemaMigration cancels one or all migrations, terminating any runnign ones as needed.",
```
Comment (ajm188) [inline go/cmd/vtctldclient/command/onlineddl.go:48]:
```suggestion
		Short:                 "CancelSchemaMigration cancels one or all migrations, terminating any running ones as needed.",
```

File: go/vt/vtctl/grpcvtctldserver/server.go [hunk 2]
```diff
@@ -513,6 +513,37 @@ func (s *VtctldServer) backupTablet(ctx context.Context, tablet *topodatapb.Tabl
...
+
+	span.Annotate("keyspace", req.Keyspace)
+	span.Annotate("uuid", req.Uuid)
+
+	query, err := alterSchemaMigrationQuery("Cancel", req.Uuid)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info("Calling ApplySchema to Cancel migration")
```
Comment (ajm188) [inline go/vt/vtctl/grpcvtctldserver/server.go:531]: i don't have an opinion _which_ way we go, but i wrote the log lines for retry/cleanup as lowercase (e.g. "Calling ApplySchema to retry migration"). we should either uppercase mine, or lowercase these for consistency, I defer to your preference on which!

File: go/vt/vtctl/grpcvtctldserver/server.go [hunk 3]
```diff
@@ -630,7 +661,6 @@ func (s *VtctldServer) CleanupSchemaMigration(ctx context.Context, req *vtctldat
 	qr, err := s.ApplySchema(ctx, &vtctldatapb.ApplySchemaRequest{
 		Keyspace:            req.Keyspace,
 		Sql:                 []string{query},
-		SkipPreflight:       true,
```
Comment (ajm188) [inline go/vt/vtctl/grpcvtctldserver/server.go:633]: ah shoot, thanks. meant to make this change in my PR!

**Review by ajm188 — APPROVED** [review]
looks good! a few minor things but approving nonetheless

File: proto/vtctlservice.proto [hunk 4]
```diff
@@ -58,6 +58,8 @@ service Vtctld {
   // This is used primarily to arrange replicas, and it will not convert a
   // primary. For that, use InitShardPrimary.
   //
+  // CancelSchemaMigration cancels one or all migrations, terminating any runnign ones as needed.
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply proto/vtctlservice.proto]: fixed

File: go/vt/vtctl/grpcvtctldserver/server.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/vtctl/grpcvtctldserver/server.go:531]: Lower case it is.

**Review by ajm188 — APPROVED** [review]

**Comment by ajm188** [comment]
looks great!!

**Review by rohit-nayak-ps — APPROVED** [review]

**Comment by shlomi-noach** [comment]
Docs PR: https://github.com/vitessio/website/pull/1572

**Pull request merged**